File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/ir/Block.java
package ro.pub.cs.diploma.ir;

import com.intellij.openapi.util.Ref;
import com.intellij.psi.PsiExpression;
import com.intellij.psi.PsiReturnStatement;
import com.intellij.psi.PsiStatement;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Block implements Statement {
  private final int id;
  @NotNull private final List<Statement> myStatements = new ArrayList<>();
  @NotNull private final List<Ref<Block>> myInBlocks = new ArrayList<>();
  @NotNull private final List<Ref<Block>> myOutBlocks = new ArrayList<>();

  private boolean doNotInline;
  private boolean finished;

  public Block(final int id) {
    this.id = id;
  }

  public void addConditionalJump(@NotNull final PsiExpression condition,
                                 @NotNull final Ref<Block> thenBlockRef,
                                 @NotNull final Ref<Block> elseBlockRef) {
    if (finished) {
      return;
    }
    myStatements.add(new ConditionalJumpStatement(condition, thenBlockRef, elseBlockRef));
    addEdgeTo(thenBlockRef);
    addEdgeTo(elseBlockRef);
    finished = true;
  }

  public void addUnconditionalJump(@NotNull final Ref<Block> blockRef) {
    if (finished) {
      return;
    }
    myStatements.add(new UnconditionalJumpStatement(blockRef));
    addEdgeTo(blockRef);
    finished = true;
  }

  public void addReturnStatement(@NotNull final PsiReturnStatement statement) {
    myStatements.add(new ReturnStatement(statement));
    finished = true;
  }

  public void addSwitchStatement(@NotNull final PsiExpression expression, @NotNull final List<Statement> statements) {
    myStatements.add(new SwitchStatement(expression, statements));
    for (Statement statement : statements) {
      if (statement instanceof UnconditionalJumpStatement) {
        addEdgeTo(((UnconditionalJumpStatement)statement).getBlockRef());
      }
    }
    finished = true;
  }

  public void add(@NotNull final PsiStatement statement) {
    myStatements.add(new NormalStatement(statement));
  }

  public void addEdgeTo(@NotNull final Ref<Block> blockRef) {
    myOutBlocks.add(blockRef);
    blockRef.get().myInBlocks.add(blockRef);
  }

  public boolean isInlinable() {
    return myInBlocks.size() == 1 && !doNotInline;
  }

  public boolean isFinished() {
    return finished;
  }

  @NotNull
  public List<Ref<Block>> getOutBlocks() {
    return myOutBlocks;
  }

  public int getId() {
    return id;
  }

  @NotNull
  List<Statement> getStatements() {
    return myStatements;
  }

  public void setDoNotInline(final boolean doNotInline) {
    this.doNotInline = doNotInline;
  }

  public boolean inlineIfTrivial() {
    if (id != 0 && myStatements.size() == 1 && myStatements.get(0) instanceof UnconditionalJumpStatement) {
      final Block jumpBlock = ((UnconditionalJumpStatement)myStatements.get(0)).getBlock();
      for (final Ref<Block> inBlock : myInBlocks) {
        inBlock.set(jumpBlock);
      }
      if (doNotInline) {
        jumpBlock.doNotInline = true;
      }
      return false;
    }
    return true;
  }

  @Override
  public void accept(@NotNull final Visitor visitor) {
    visitor.visit(this);
  }

  public List<String> toDot() {
    List<String> strings = new ArrayList<>();
    if (myStatements.size() > 0) {
      for (Statement statement : myStatements.subList(0, myStatements.size() - 1)) {
        strings.add(((WrapperStatement)statement).getStatement().getText()
                      .replace("{", "\\{")
                      .replace("}", "\\}")
                      .replace("<", "\\<")
                      .replace(">", "\\>")
                      .replace("\n", "\\n"));
      }
    }

    List<String> strings2 = new ArrayList<>();
    strings2.add(String.format("id: %s", id));
    strings2.add(strings.stream().collect(Collectors.joining("\\n")));
    String color = doNotInline ? "color=red" : "";
    List<String> statements = new ArrayList<>();
    if (myStatements.size() > 0) {
      final Statement lastStatement = myStatements.get(myStatements.size() - 1);
      if (lastStatement instanceof ConditionalJumpStatement) {
        strings2.add(((ConditionalJumpStatement)lastStatement).getCondition().getText()
                         .replace("{", "\\{")
                         .replace("}", "\\}")
                         .replace("<", "\\<")
                         .replace(">", "\\>")
                         .replace("\n", "\\n"));
        strings2.add("{<true>true|<false>false}");
        statements.add(String.format("%s:true -> %s;", id, ((ConditionalJumpStatement)lastStatement).getThenBlock().getId()));
        statements.add(String.format("%s:false -> %s;", id, ((ConditionalJumpStatement)lastStatement).getElseBlock().getId()));
      }
      else if (lastStatement instanceof UnconditionalJumpStatement) {
        final int jumpId = ((UnconditionalJumpStatement)lastStatement).getBlock().getId();
        strings2.add(String.format("jump %s;", jumpId));
        statements.add(String.format("%s -> %s;", id, jumpId));
      } else if (lastStatement instanceof ReturnStatement) {
        strings2.add(((ReturnStatement)lastStatement).getStatement().getText()
                       .replace("{", "\\{")
                       .replace("}", "\\}")
                       .replace("<", "\\<")
                       .replace(">", "\\>")
                       .replace("\n", "\\n"));
      }
    }
    statements.add(String.format("%s [label=\"{%s}\" %s];", id, strings2.stream().collect(Collectors.joining("|")), color));

    return statements;
  }
}

File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/ir/Block.kt
package ro.pub.cs.diploma.ir

import com.intellij.openapi.util.Ref
import com.intellij.psi.PsiExpression
import com.intellij.psi.PsiReturnStatement
import com.intellij.psi.PsiStatement

class Block(val id: Int) : Statement {
  private val myStatements = ArrayList<Statement>()
  private val myInBlocks = ArrayList<Ref<Block>>()
  private val myOutBlocks = ArrayList<Ref<Block>>()

  private var doNotInline: Boolean = false
  var isFinished: Boolean = false
    private set

  val isInlinable: Boolean
    get() = myInBlocks.size == 1 && !doNotInline

  val outBlocks: List<Ref<Block>>
    get() = myOutBlocks

  internal val statements: List<Statement>
    get() = myStatements

  fun addConditionalJump(condition: PsiExpression,
                         thenBlockRef: Ref<Block>,
                         elseBlockRef: Ref<Block>) {
    if (isFinished) {
      return
    }
    myStatements.add(ConditionalJumpStatement(condition, thenBlockRef, elseBlockRef))
    addEdgeTo(thenBlockRef)
    addEdgeTo(elseBlockRef)
    isFinished = true
  }

  fun addUnconditionalJump(blockRef: Ref<Block>) {
    if (isFinished) {
      return
    }
    myStatements.add(UnconditionalJumpStatement(blockRef))
    addEdgeTo(blockRef)
    isFinished = true
  }

  fun addReturnStatement(statement: PsiReturnStatement) {
    myStatements.add(ReturnStatement(statement))
    isFinished = true
  }

  fun addSwitchStatement(expression: PsiExpression, statements: List<Statement>) {
    myStatements.add(SwitchStatement(expression, statements))
    statements.filterIsInstance<UnconditionalJumpStatement>().forEach { addEdgeTo(it.blockRef) }
    isFinished = true
  }

  fun add(statement: PsiStatement) {
    myStatements.add(NormalStatement(statement))
  }

  fun addEdgeTo(blockRef: Ref<Block>) {
    myOutBlocks.add(blockRef)
    blockRef.get().myInBlocks.add(blockRef)
  }

  fun setDoNotInline(doNotInline: Boolean) {
    this.doNotInline = doNotInline
  }

  fun inlineIfTrivial(): Boolean {
    if (id != 0 && myStatements.size == 1 && myStatements[0] is UnconditionalJumpStatement) {
      val jumpBlock = (myStatements[0] as UnconditionalJumpStatement).block
      for (inBlock in myInBlocks) {
        inBlock.set(jumpBlock)
      }
      if (doNotInline) {
        jumpBlock.doNotInline = true
      }
      return false
    }
    return true
  }

  override fun accept(visitor: Visitor) {
    visitor.visit(this)
  }

  fun toDot(): List<String> {
    val strings = ArrayList<String>()
    if (myStatements.size > 0) {
      myStatements.subList(0, myStatements.size - 1).mapTo(strings) {
        (it as WrapperStatement).statement.text
            .replace("{", "\\{")
            .replace("}", "\\}")
            .replace("<", "\\<")
            .replace(">", "\\>")
            .replace("\n", "\\n")
      }
    }

    val strings2 = ArrayList<String>()
    strings2.add("id: $id")
    strings2.add(strings.joinToString("\\n"))
    val statements = ArrayList<String>()
    if (myStatements.size > 0) {
      val lastStatement = myStatements.last()
      when (lastStatement) {
        is ConditionalJumpStatement -> {
          strings2.add(lastStatement.condition.text
              .replace("{", "\\{")
              .replace("}", "\\}")
              .replace("<", "\\<")
              .replace(">", "\\>")
              .replace("\n", "\\n"))
          strings2.add("{<true>true|<false>false}")
          statements.add("$id:true -> ${lastStatement.thenBlock.id};")
          statements.add("$id:false -> ${lastStatement.elseBlock.id};")
        }
        is UnconditionalJumpStatement -> {
          val jumpId = lastStatement.block.id
          strings2.add("jump $jumpId;")
          statements.add("$id -> $jumpId;")
        }
        is ReturnStatement -> strings2.add(lastStatement.statement.text
            .replace("{", "\\{")
            .replace("}", "\\}")
            .replace("<", "\\<")
            .replace(">", "\\>")
            .replace("\n", "\\n"))
      }
    }
    statements.add("$id [label=\"{${strings2.joinToString("|")}}\" ${if (isInlinable) "" else "color=red"}];")

    return statements
  }
}
